File: packages/adapter-tests/vectors/probe/runners/probe.rb
# frozen_string_literal: true
# Probe runner (Ruby / ERB evaluator). Reads $PROBE_VECTORS, evaluates each
# case, and prints a classified line per divergence. Not a committed vector
# test — see ../README.md.
require 'json'
require 'barefoot_js/evaluator'

def value_match?(got, expect)
  return got.nil? if expect.nil?
  if expect.is_a?(Hash) && expect.key?(:$num)
    kind = expect[:$num]
    return false unless got.is_a?(Numeric)
    f = got.to_f
    return f.nan? if kind == 'NaN'
    return f.infinite? == (kind == 'Infinity' ? 1 : -1)
  end
  if expect == true || expect == false
    return (got == true || got == false) && got == expect
  end
  if expect.is_a?(Array)
    return false unless got.is_a?(Array) && got.length == expect.length
    return got.each_index.all? { |i| value_match?(got[i], expect[i]) }
  end
  if expect.is_a?(Hash)
    return false unless got.is_a?(Hash) && got.keys.length == expect.keys.length
    return expect.keys.all? { |k| got.key?(k) && value_match?(got[k], expect[k]) }
  end
  return false if got.nil?
  return false if got.is_a?(Numeric) != expect.is_a?(Numeric)
  return got == expect if got.is_a?(Numeric)
  return false unless got.is_a?(String) && expect.is_a?(String)
  got == expect
end

cases = JSON.parse(File.read(ENV['PROBE_VECTORS'], encoding: Encoding::UTF_8), symbolize_names: true)[:cases]
cases.each do |c|
  begin
    got = BarefootJS::Evaluator.evaluate(c[:expr], c[:env] || {})
    unless value_match?(got, c[:expect])
      kind = c[:known] ? 'KNOWN' : 'NEW'
      puts "#{kind}\t#{c[:category]}\t#{c[:note]}\t#{got.inspect}\t#{c[:expect].inspect}"
    end
  rescue => e
    puts "ERROR\t#{c[:category]}\t#{c[:note]}\t#{e.class}: #{e.message}"
  end
end
puts "RAN\t#{cases.length}"

File: packages/adapter-tests/vectors/probe/runners/probe.php
<?php
declare(strict_types=1);
// Probe runner (PHP evaluator). Reads $PROBE_VECTORS. See ../README.md.
require_once __DIR__ . '/../../../../adapter-php/src/Evaluator.php';
use Barefoot\Evaluator;

function bfe_match($got, $expect): bool {
    if ($expect === null) return $got === null;
    if (is_array($expect) && array_key_exists('$num', $expect) && count($expect) === 1) {
        $kind = $expect['$num'];
        if (is_bool($got) || !(is_int($got) || is_float($got))) return false;
        $g = (float) $got;
        if ($kind === 'NaN') return is_nan($g);
        return $g === ($kind === 'Infinity' ? INF : -INF);
    }
    if (is_bool($expect)) return is_bool($got) && $got === $expect;
    if (is_array($expect) && array_is_list($expect)) {
        if (!is_array($got) || !array_is_list($got) || count($got) !== count($expect)) return false;
        foreach ($expect as $i => $e) if (!bfe_match($got[$i] ?? null, $e)) return false;
        return true;
    }
    if (is_array($expect)) {
        $gotArr = $got instanceof \stdClass ? get_object_vars($got) : (is_array($got) ? $got : null);
        if ($gotArr === null || count($gotArr) !== count($expect)) return false;
        foreach ($expect as $k => $v) if (!array_key_exists($k, $gotArr) || !bfe_match($gotArr[$k], $v)) return false;
        return true;
    }
    if ($got === null || is_array($got) || $got instanceof \stdClass) return false;
    $wantNum = is_int($expect) || is_float($expect);
    $gotNum = is_int($got) || is_float($got);
    if ($wantNum !== $gotNum) return false;
    if ($wantNum) return (float) $got === (float) $expect;
    return $got === $expect;
}
function fmt($v): string {
    if (is_float($v) && (is_nan($v) || is_infinite($v))) return (string) $v;
    $j = json_encode($v);
    return $j === false ? var_export($v, true) : $j;
}

$doc = json_decode(file_get_contents(getenv('PROBE_VECTORS')), true);
$n = 0;
foreach ($doc['cases'] as $c) {
    $n++;
    try {
        $got = Evaluator::evaluate($c['expr'], $c['env'] ?? []);
    } catch (\Throwable $e) {
        echo "ERROR\t{$c['category']}\t{$c['note']}\t" . get_class($e) . ": " . $e->getMessage() . "\n";
        continue;
    }
    if (!bfe_match($got, $c['expect'])) {
        $kind = !empty($c['known']) ? 'KNOWN' : 'NEW';
        echo "{$kind}\t{$c['category']}\t{$c['note']}\t" . fmt($got) . "\t" . fmt($c['expect']) . "\n";
    }
}
echo "RAN\t{$n}\n";
